pystac/rel_type.py

```diff
@@ -0,0 +1,33 @@
+from enum import Enum
+
+
+class RelType(str, Enum):
+    """A list of common rel types that can be used in STAC Link metadata."""
+
+    def __str__(self) -> str:
+        return str(self.value)
+
+    # Core + best practices
+    ALTERNATE = "alternate"
+    CANONICAL = "canonical"
+    CHILD = "child"
+    COLLECTION = "collection"
+    ITEM = "item"
+    ITEMS = "items"
+    LICENSE = "license"
+    DERIVED_FROM = "derived_from"
+    NEXT = "next"
+    PARENT = "parent"
+    PREV = "prev"
+    PREVIEW = "preview"
+    ROOT = "root"
+    SELF = "self"
+    VIA = "via"
+
+    # Label Extension
+    SOURCE = "source"
+
+    # Version Extension
+    LATEST = "latest-version"
+    PREDECESSOR = "predecessor-version"
+    SUCCESSOR = "successor-version"
```

Comment:
Including extension reltypes makes sense to me, though I could see a situation where someone is confused about using a reltype outside of the extension context or expecting to find the reltype in the extension. I'd say I'm +0 on this approach.

Reply:
To be honest, the more I sit with this approach the less I like it. I think it might be clearer to have rel types defined by extensions sitting with that extension code. 

Here's an alternative proposal:

1) Create a base `RelType` enum:

    ```python
    class RelType(str, Enum):
        def __str__(self) -> str:
            return str(self.value)
    ``` 

2) Have a `CoreRelType` enum that actually defines the values from the core spec and best practices:

    ```python
    class CoreRelType(RelType):
        ALTERNATE = "alternate"
        ...
    ```

3) Extension define their own sub-class as needed:

    ```python
    class LabelExtensionRelType(RelType):
        SOURCE = "source"
    ```

The type annotations for all the functions/classes that use rel types would stay as `rel: Union[str, RelType]` and extended rel types would be more clearly associated with the right extension. The downside would be that you wouldn't have a single place to enumerate all STAC rel types, you would need to specifically import something like `pystac.extensions.label.LabelExtensionRelType`, but maybe that explicit import is better anyway...

Reply:
Yeah that makes sense to me. A little more work to use but a clearer organization.

I do think that having `RelType` hold all of the core rel types instead of making a separate `CoreRelType` would make sense; that would make LabelExtensionRelType also hold those core reltypes which is weird but it's probably more convenient to have the base RelType have the core instead of having to figure out there's a "core". Not sure though.

`LabelExtension` is getting a bit long, maybe `LabelRelType` or `LabelExtRelType`? Or perhaps it's best to have the full name for clarity, I usually don't mind additional typing if it makes code more consistent and clear. I do sometimes get accused of writing enterprise-y looking code though with long names 😆

Reply:
Python [only allows you to subclass an enum with no members](https://docs.python.org/3/library/enum.html#restricted-enum-subclassing), so we wouldn't be able to have `LabelExtRelType` inherit from `RelType` if it included the core members.

It looks like we don't need to, though. Since it inherits from `str`, `LabelExtRelType` will pass a type check against `str` in any function arguments, so it doesn't need to inherit from `RelType`. As you mentioned in an earlier comment, we're really just using the `Union[str, RelType]` annotation to let users know they should use `RelType` if possible. 

Given all of that, I would be for removing the extension-specific rel types from `pystac.rel_type.RelType` and putting them into their own enums in the extension modules.

Reply:
I implemented this in 2d53f0b1b4f5bca32bc3b3e96facb38f8a66c706 so we can see how it feels. I'll revert it if we decide to go a different route.